Interface for determining package pins

robtaylor opened the pull request:
# Package Pin Interface

This PR implements a structured interface for determining package pins in ChipFlow.

## Key Changes

1. Added abstract methods to define standard pin types in package definitions:
   - Power and ground pins
   - Clock pins
   - Reset pins
   - JTAG pins
   - Heartbeat pins

2. Implemented these methods in both package types:
   - BareDiePackageDef
   - QuadPackageDef

3. Enhanced the Package class to:
   - Support new pin types
   - Initialize pins from package definitions
   - Support both legacy and new pin formats

4. Added comprehensive tests in test_package_pins.py
5. Created documentation in docs/package_pins.md

## Backward Compatibility

The implementation maintains backward compatibility with the existing TOML configuration format while introducing a new, more flexible format.

## Ready for Review

All tests are passing, and the implementation is ready for review.

gatecat commented on chipflow_lib/platforms/utils.py:
```diff
@@ -114,6 +114,16 @@ def BidirPinSignature(width, **kwargs):
 PinList = List[Pin]
 Pins = Union[PinSet, PinList]
 
+class PowerPins(enum.Enum):
```
We probably at least want to split IO vs core here

robtaylor replied:
good point. what sort of information does the IO power pins needs?

gatecat replied:
I think this is what Serge brought up on Element already a few days ago. Basically right now we have all the IO in one bank with no breaker cells used, so we just have IOVdd/IOVss that apply to all IO pins. But eventually we want to have multiple groups of IO pins (potentially including analog) in seperate banks, each associated with its own one or more IO power and ground pins.

robtaylor replied:
@lanserge  @gatecat I'm thinking how best to do this. is it a case that a port should have a request for an io power bank? and the pin allocation then should be able to have rules for where on the padring to allocate that?

robtaylor replied:
Maybe we introduce the idea of a power domain? Might be too much to resolve properly at this point without concrete use cases :/

lanserge replied:
Power domain could be good idea, but domains could have relations, or even connected inside.

stafverhaegen-chipflow replied:
I think this has to be further split up in different domains for core and different domains for IO. You will have breaker IO cells that will break the core and/or the IO supply/ground rings in the IO ring.
Multiple core power domains will also need support from Amaranth to know which logic has to be in which power domain.
An anlog input does not necessarily need to be in separate IO power domain but typically one wants to have a separate power domain for the analog block connected to the analog input to isolate analog block from the digital switching noise on the vdd/gnd signals from the core logic.

lanserge commented on chipflow_lib/platforms/utils.py:
```diff
@@ -206,18 +216,70 @@ class _BasePackageDef(pydantic.BaseModel, abc.ABC):
+
+    @property
+    @abc.abstractmethod
+    def heartbeat(self) -> Dict(int, Pin):
```
The clock in some cases comes as two pins close to each other to be connected to XTAL.

robtaylor replied:
hmm, good point. Is that a config option to generate a oscillator?

lanserge replied:
Currently we have nothing to specify use of XTAL, this needs to be added somehow.

stafverhaegen-chipflow replied:
> The clock in some cases comes as two pins close to each other to be connected to XTAL.

If you have external crystal you will need a crystal oscillator analog block on the chip. The actual clock will then be an output of this oscillator.

stafverhaegen-chipflow replied:
What also can happen is that clock is delivered as differential signal; but this can also be for digital signals. Likely only to add when use case is there.

robtaylor commented:
I've updated this PR to implement a complete interface for package pins as discussed. 

Key enhancements:

1. Implemented all abstract methods in both package types with sensible defaults
2. Added support for new power pin format while maintaining backward compatibility
3. Created comprehensive tests with 76% code coverage
4. Added documentation explaining the new interface

All tests are passing. Ready for review\!

gatecat commented on docs/chipflow-toml-guide.rst:
```diff
@@ -139,19 +140,37 @@ Available pad rings
+The package definition provides default locations for standard pins like power, ground, clocks, and resets. You only need to specify the name and properties.
+
+[chipflow.silicon.power]
+vdd = { type = "power", name = "vdd", voltage = "1.8V" }
```
it'd be nice to have an example of separate core and IO power here, as that will always almost be the case

gatecat commented on docs/chipflow-toml-guide.rst:
```diff
@@ -139,19 +140,37 @@ Available pad rings
-   sys_rst_n = { type = "reset", loc = "115" }
 
-In the above example two pads specified, ``sys_clk`` pad for clock input and ``sys_rst_n`` for reset.
+``[silicon.power]``
```
section is more-or-less repeated?

gatecat reviewed (commented):
Haven't looked through the tests yet, but these are my comments so far

robtaylor commented on chipflow_lib/config_models.py:
```diff
@@ -36,13 +40,16 @@ def validate_pad_dict(cls, v: dict, info: ValidationInfo):
     package: Literal["caravel", "cf20", "pga144"]
-    pads: Dict[str, PadConfig] = {}
-    power: Dict[str, PadConfig] = {}
+    power: Dict[str, Voltage] = {}
```
I think the validation will error for any old configis

robtaylor commented:
Has some WIP for better pin allocation.

This should become shell and padring aware
